test/with_dummyserver/test_socketlevel.py

```diff
@@ -1014,6 +1014,40 @@ def consume_ssl_socket(listener: socket.socket) -> None:
             ssl_sock.sendall(b"hello")
             assert ssl_sock.fileno() > 0
 
+    def test_socket_shutdown_stops_recv(self) -> None:
+        timed_out = Event()
+
+        def socket_handler(listener: socket.socket) -> None:
+            sock = listener.accept()[0]
+
+            # Consume request
+            buf = b""
+            while not buf.endswith(b"\r\n\r\n"):
+                buf = sock.recv(65535)
+
+            # Send incomplete message (note Content-Length)
+            sock.send(
+                b"HTTP/1.1 200 OK\r\n"
+                b"Content-Type: text/plain\r\n"
+                b"Content-Length: 10\r\n"
+                b"\r\n"
+                b"Hi-"
+            )
+            timed_out.wait(5)
+            sock.close()
+
+        self._start_server(socket_handler)
+
+        with HTTPConnectionPool(self.host, self.port) as pool:
+            response = pool.urlopen("GET", "/", preload_content=False, retries=0)
+            # Calling shutdown here calls shutdown() on the underlying socket,
+            # so that the reamining read will fail instead of blocking
+            # indefinitely
+            response.shutdown()
```

Comment:
Calling `response.close()` here instead of `response.shutdown()` doesn't make the read call hang
Would it need to block in case of calling `close` for correct testing?

Reply:
The read() call does hang, actually.

The *test* itself does not hang, because after 5 seconds, the server will call `sock.close()` itself. But in that case we don't get any data, so no `ProtocolError`, simply the empty string, and the test will fail as expected.

I prefer failing after 5 seconds than hanging, because hanging tests are annoying to debug.

Reply:
It didn't wait for five seconds with `response.close()`, read returned right away, the following change printed 29980.729741648 and 29980.729781285

```diff
diff --git a/test/with_dummyserver/test_socketlevel.py b/test/with_dummyserver/test_socketlevel.py
index 46feb4bf..ba45a2b5 100644
--- a/test/with_dummyserver/test_socketlevel.py
+++ b/test/with_dummyserver/test_socketlevel.py
@@ -14,6 +14,7 @@ import socket
 import ssl
 import tempfile
 import threading
+import time
 import typing
 import zlib
 from collections import OrderedDict
@@ -1043,9 +1044,11 @@ class TestSocketClosing(SocketDummyServerTestCase):
             # Calling shutdown here calls shutdown() on the underlying socket,
             # so that the remaining read will fail instead of blocking
             # indefinitely
-            response.shutdown()
-            with pytest.raises(ProtocolError, match="Connection broken"):
-                response.read()
+            response.close()
+            # with pytest.raises(ProtocolError, match="Connection broken"):
+            print(time.monotonic())
+            response.read()
+            print(time.monotonic())
             timed_out.set()


```

No protocol error raised indeed

Reply:
Right, the test as a whole takes 5 seconds, but only because the teardown only completes when `socket_handler` returns. I initially had two client threads to recreate the original report, and they did not seem to be needed.

They are in fact needed, so I'll try to add that back to reproduce the initial report better, thanks.